pkg/exchange/coinbase/stream.go

```diff
@@ -21,10 +33,146 @@ type Stream struct {
 	activeMessageCallbacks   []func(m *ActiveMessage)
 }
 
-func NewStream() *Stream {
+func NewStream(
+	apiKey string,
+	passphrase string,
+	secretKey string,
+) *Stream {
 	s := Stream{
 		StandardStream: types.NewStandardStream(),
+		apiKey:         apiKey,
+		passphrase:     passphrase,
+		secretKey:      secretKey,
 	}
 	s.SetParser(s.parseMessage)
+	s.SetDispatcher(s.dispatchEvent)
+	s.SetEndpointCreator(createEndpoint)
+
+	// public handlers
+	s.OnConnect(s.handleConnect)
+
 	return &s
 }
+
+func (s *Stream) dispatchEvent(e interface{}) {
+	switch e := e.(type) {
+	case *StatusMessage:
+		s.EmitStatusMessage(e)
+	case *AuctionMessage:
+		s.EmitAuctionMessage(e)
+	case *RfqMessage:
+		s.EmitRfqMessage(e)
+	case *TickerMessage:
+		s.EmitTickerMessage(e)
+	case *ReceivedMessage:
+		s.EmitReceivedMessage(e)
+	case *OpenMessage:
+		s.EmitOpenMessage(e)
+	case *DoneMessage:
+		s.EmitDoneMessage(e)
+	case *MatchMessage:
+		s.EmitMatchMessage(e)
+	case *ChangeMessage:
+		s.EmitChangeMessage(e)
+	case *ActiveMessage:
+		s.EmitActiveMessage(e)
+	default:
+		log.Warnf("skip dispatching msg due to unknown message type: %T", e)
+	}
+}
+
+func createEndpoint(ctx context.Context) (string, error) {
+	return wsFeedURL, nil
+}
+
+type channelType struct {
+	Name       string   `json:"name"`
+	ProductIDs []string `json:"product_ids,omitempty"`
+}
+
+type websocketCommand struct {
+	Type       string        `json:"type"`
+	Channels   []channelType `json:"channels"`
+	Signature  *string       `json:"signature,omitempty"`
+	Key        *string       `json:"key,omitempty"`
+	Passphrase *string       `json:"passphrase,omitempty"`
+	Timestamp  *string       `json:"timestamp,omitempty"`
+}
+
+func (s *Stream) handleConnect() {
+	// subscribe to channels
+	if len(s.Subscriptions) == 0 {
+		return
+	}
+
+	subProductsMap := make(map[string][]string)
+	for _, sub := range s.Subscriptions {
+		strChannel := string(sub.Channel)
+		// "rfq_matches" allow empty symbol
+		if sub.Channel != "rfq_matches" && len(sub.Symbol) == 0 {
+			continue
+		}
+		subProductsMap[strChannel] = append(subProductsMap[strChannel], sub.Symbol)
+	}
+	subCmds := []websocketCommand{}
+	signature, ts := s.generateSignature()
+	for channel, productIDs := range subProductsMap {
+		var subType string
+		switch channel {
+		case "rfq_matches":
+			subType = "subscriptions"
+		default:
+			subType = "subscribe"
+		}
+		subCmd := websocketCommand{
+			Type: subType,
+			Channels: []channelType{
+				{
+					Name:       channel,
+					ProductIDs: productIDs,
+				},
+			},
+		}
+		if s.authEnabled() {
+			subCmd.Signature = &signature
+			subCmd.Key = &s.apiKey
+			subCmd.Passphrase = &s.passphrase
+			subCmd.Timestamp = &ts
+		}
+		subCmds = append(subCmds, subCmd)
+	}
+	for _, subCmd := range subCmds {
+		err := s.Conn.WriteJSON(subCmd)
+		if err != nil {
+			log.WithError(err).Errorf("subscription error: %v", subCmd)
```

Comment:
Should we return on error?

Reply:
It seems to me that websocket subscription is not of all-or-nothing semantic.
The user can still work on the feeds of the successful subscriptions despite that there are failed ones.